src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -111,163 +453,135 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
+    } else {
+      log_trace(gc, free)("Failed to shrink TLAB or GCLAB request (" SIZE_FORMAT ") in region " SIZE_FORMAT " to " SIZE_FORMAT
+                          " because min_size() is " SIZE_FORMAT, req.size(), r->index(), adjusted_size, req.min_size());
     }
   } else {
+    size_t size = req.size();
     result = r->allocate(size, req.type());
+    if (result != nullptr) {
+      // Record actual allocation size
+      log_debug(gc, free)("Allocated " SIZE_FORMAT " words for %s @" PTR_FORMAT
+                          " from %s region " SIZE_FORMAT ", free bytes remaining: " SIZE_FORMAT,
+                          size, ShenandoahAllocRequest::alloc_type_to_string(req.type()), p2i(result),
+                          partition_name(_partitions.membership(r->index())),  r->index(), r->free());
+      req.set_actual_size(size);
+    }
   }
 
   if (result != nullptr) {
     // Allocation successful, bump stats:
     if (req.is_mutator_alloc()) {
-      increase_used(size * HeapWordSize);
-    }
-
-    // Record actual allocation size
-    req.set_actual_size(size);
+      _partitions.increase_used(Mutator, req.actual_size() * HeapWordSize);
+    } else {
+      assert(req.is_gc_alloc(), "Should be gc_alloc since req wasn't mutator alloc");
 
-    if (req.is_gc_alloc()) {
+      // For GC allocations, we advance update_watermark because the objects relocated into this memory during
+      // evacuation are not updated during evacuation.
       r->set_update_watermark(r->top());
     }
   }
 
-  if (result == nullptr || has_no_alloc_capacity(r)) {
-    // Region cannot afford this or future allocations. Retire it.
-    //
-    // While this seems a bit harsh, especially in the case when this large allocation does not
-    // fit, but the next small one would, we are risking to inflate scan times when lots of
-    // almost-full regions precede the fully-empty region where we want allocate the entire TLAB.
-    // TODO: Record first fully-empty region, and use that for large allocations
+  if (alloc_capacity(r) < PLAB::min_size() * HeapWordSize) {
+    // Regardless of whether this allocation succeeded, if the remaining memory is less than PLAB:min_size(), retire this region.
+    // Note that retire_within_partition() increases used to account for waste.
 
-    // Record the remainder as allocation waste
-    if (req.is_mutator_alloc()) {
-      size_t waste = r->free();
-      if (waste > 0) {
-        increase_used(waste);
-        _heap->notify_mutator_alloc_words(waste >> LogHeapWordSize, true);
-      }
-    }
+    // Note that a previous implementation of this function would retire a region following any failure to
+    // allocate within.  This was observed to result in large amounts of available memory being ignored
+    // following a failed shared allocation request.  TLAB requests will generally downsize to absorb all
+    // memory available within the region even if this is less than the desired size.
```

Comment:
I don't understand this comment, since you are it seems to me retiring the region below at line 553. (Also see comment elsewhere on calling the method `retire_within_partition`, instead of the more natural (to me) `retire_from_partition`.

Reply:
I'm fixing this comment to make more clear.  In the current implementation, we only retire a region if the remaining capacity is less than PLAB::min_size().  The previous implementation was observed to retire some regions even when there was 50% of the region still available (in the case that a very large shared alloc failed).